File: src/main/java/com/handwoong/everyonewaiter/domain/Category.java
package com.handwoong.everyonewaiter.domain;

import static jakarta.persistence.FetchType.LAZY;
import static lombok.AccessLevel.PRIVATE;
import static lombok.AccessLevel.PROTECTED;

import com.handwoong.everyonewaiter.dto.CategoryDto;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import java.util.ArrayList;
import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = PROTECTED)
public class Category extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "category_id")
    private Long id;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "member_id")
    private Store store;

    private String name;

    @OneToMany(mappedBy = "category", cascade = CascadeType.ALL)
    private final List<Menu> menuList = new ArrayList<>();

    @Builder(access = PRIVATE)
    private Category(Store store, String name) {
        this.store = store;
        this.name = name;
    }

    public static Category createCategory(Store store, CategoryDto.RequestDto categoryDto) {
        return Category.builder()
                .store(store)
                .name(categoryDto.getName())
                .build();
    }

    public void update(CategoryDto.RequestDto categoryDto) {
        this.name = categoryDto.getName();
    }

    public void addMenu(Menu menu) {
        getMenuList().add(menu);
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/domain/category/Category.kt
package com.handwoong.everyonewaiter.domain.category

import com.handwoong.everyonewaiter.domain.BaseEntity
import com.handwoong.everyonewaiter.domain.menu.Menu
import com.handwoong.everyonewaiter.domain.store.Store
import com.handwoong.everyonewaiter.dto.category.CategoryRequest
import javax.persistence.*
import javax.persistence.FetchType.LAZY
import javax.persistence.GenerationType.IDENTITY

@Entity
class Category(

    var name: String,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "member_id")
    val store: Store,

    @OneToMany(mappedBy = "category", cascade = [CascadeType.ALL])
    val menuList: MutableList<Menu> = mutableListOf(),

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "category_id")
    val id: Long? = null,
) : BaseEntity() {

    fun update(categoryDto: CategoryRequest) {
        this.name = categoryDto.name
    }

    fun addMenu(menu: Menu) {
        this.menuList.add(menu)
    }

    companion object {
        fun createCategory(
            categoryDto: CategoryRequest,
            store: Store,
        ): Category {
            return Category(
                name = categoryDto.name,
                store = store,
            )
        }
    }

}
